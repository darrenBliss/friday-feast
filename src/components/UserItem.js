// @flow
import * as React from 'react'
import styled from 'styled-components'

const Holder = styled.div`
  background-color: #fff;
  padding: 1rem;
  border-bottom: 1px solid #eee;
  background-color: ${props => props.isActive && 'papayawhip'};
`

const ItemText = styled.span`
  display: inline-block;
  padding: 1rem;
  margin-right: 1rem;
`

const Button = styled.button`
  display: inline-block;
  font-size: 1rem;
  color: #fff;
  padding: 1rem;
  margin: .2rem;
  background-color: #FF5A5F;
  border: 0;
  border-radius: 3px;
`

type Props = {
  name: string,
  uid: number,
  index: number,
  isActive: ?boolean,
  onRemove: (number) => void,
  userEdited: (string, number) => void,
  onMove: (?number, number) => void
}

type State = {
  isEdit: boolean,
  name: string
}

class UserItem extends React.Component<Props, State> {
  myInput: ?HTMLInputElement
  removeHandler: () => void
  triggerEdit: () => void
  cancelEdit: () => void
  handleSet: () => void
  onNameChange: () => void
  move: (number) => void

  constructor(props: Props) {
    super(props)

    this.state = {
      isEdit: false,
      name: this.props.name
    }

    this.removeHandler = this.removeHandler.bind(this)
    this.triggerEdit = this.triggerEdit.bind(this)
    this.cancelEdit = this.cancelEdit.bind(this)
    this.handleSet = this.handleSet.bind(this)
    this.onNameChange = this.onNameChange.bind(this)
    // this.move = this.move.bind(this)
  }

  removeHandler() {
    const { uid } = this.props
    this.props.onRemove(uid)
  }

  triggerEdit() {
    this.setState({ isEdit: true }, () => {
      if ( this.myInput ) {
        this.myInput.focus()
      }
    })
  }

  cancelEdit() {
    // console.log('cancel edit')
    this.setState({
      isEdit: false,
      name: this.props.name
    })
  }

  handleSet(event: SyntheticEvent<>) {
    event.preventDefault()

    this.props.userEdited(this.state.name, this.props.uid)
    this.setState({
      isEdit: false
    })
  }

  onNameChange(event: SyntheticInputEvent<HTMLInputElement>) {
    this.setState({
      name: event.currentTarget.value
    })
  }

  move(dir: number) {
    this.props.onMove(dir, this.props.index)
  }

  render() {
    return (
      <Holder isActive={this.props.isActive && true}>
        <button type="button" data-dir="-1" onClick={this.move.bind(this,-1)}>⬆️ Up</button>
        <button type="button" data-dir="1" onClick={this.move.bind(this,1)}>⬇️ Down</button>
        { this.state.isEdit ?
          (
            <div>
              <form onSubmit={this.handleSet}>
                <input
                  ref={input => {this.myInput = input}}
                  value={this.state.name}
                  onChange={this.onNameChange}
                  type="text"/>
                <Button type="submit">Set</Button>
              </form>
              <Button type="button" onClick={this.cancelEdit}>Cancel</Button>
            </div>
          )
          :
          (
            <div>
              <ItemText>{ this.props.name }</ItemText>
              <Button type="button" onClick={this.triggerEdit}>Edit</Button>
              <Button type="button" onClick={this.removeHandler}>Delete</Button>
            </div>
          )
        }
      </Holder>
    )
  }
}

export default UserItem