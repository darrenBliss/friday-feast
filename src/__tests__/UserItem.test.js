import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import UserItem from '../components/UserItem'


describe('UserItem component', () => {
  it('Render snapshot', () => {
    const wrapper = shallow(<UserItem
      name="Megatron"
      index="8"
      uid="7"
      isActive="true"
    />)

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('Component active state', () => {
    const wrapper = shallow(<UserItem
      isActive="true"
    />)

    expect(wrapper.props().isActive).toEqual(true)
  })

  it('Displays usertext at start', () => {
    const wrapper = shallow(<UserItem
      name="Megatron"
    />)

    expect(wrapper.find('UserItem__ListBox')).toHaveLength(1)
  })

  it('Displays edit form when in edit mode', () => {
    const wrapper = shallow(<UserItem
      name="Megatron"
    />)

    wrapper.find('UserItem__ListBox').find('Button').at(0).simulate('click')
    expect(wrapper.find('UserItem__EditForm')).toHaveLength(1)
  })




})

