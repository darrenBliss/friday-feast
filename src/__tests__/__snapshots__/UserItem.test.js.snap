// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserItem component Render snapshot 1`] = `
<UserItem__Holder
  isActive={true}
>
  <UserItem__DirectionButtons>
    <Button
      data-dir="-1"
      onClick={[Function]}
      title="Up"
      type="button"
    >
      ↑
    </Button>
    <Button
      data-dir="1"
      onClick={[Function]}
      title="Down"
      type="button"
    >
      ↓
    </Button>
  </UserItem__DirectionButtons>
  <UserItem__ListBox>
    <UserItem__ItemText>
      Megatron
    </UserItem__ItemText>
    <Button
      onClick={[Function]}
      type="button"
    >
      Edit
    </Button>
    <Button
      onClick={[Function]}
      type="button"
    >
      Delete
    </Button>
    <Button
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      Set as active
    </Button>
  </UserItem__ListBox>
</UserItem__Holder>
`;
